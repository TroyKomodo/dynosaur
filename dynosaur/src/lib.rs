#![doc = include_str!("../README.md")]

pub use dynosaur_derive::*;
