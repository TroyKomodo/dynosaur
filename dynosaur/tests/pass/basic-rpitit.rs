use dynosaur::dynosaur;

#[dynosaur(DynMyTrait)]
trait MyTrait {
    fn foo(&self) -> impl Send;
}

fn main() {}
