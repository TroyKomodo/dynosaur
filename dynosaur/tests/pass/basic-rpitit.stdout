#![feature(prelude_import)]
#[prelude_import]
use std::prelude::rust_2021::*;
#[macro_use]
extern crate std;
use dynosaur::dynosaur;

trait MyTrait {
    fn foo(&self)
    -> impl Send;
}
mod _dynosaur_macro_DynMyTrait {
    use super::*;
    trait ErasedMyTrait {
        fn foo<'life0, 'dynosaur>(&'life0 self)
        -> ::core::pin::Pin<Box<dyn Send + 'dynosaur>>
        where
        'life0: 'dynosaur,
        Self: 'dynosaur;
    }
    impl<DYNOSAUR: MyTrait> ErasedMyTrait for DYNOSAUR {
        fn foo<'life0, 'dynosaur>(&'life0 self)
            -> ::core::pin::Pin<Box<dyn Send + 'dynosaur>> where
            'life0: 'dynosaur, Self: 'dynosaur {
            Box::pin(<Self as MyTrait>::foo(self))
        }
    }
    pub struct DynMyTrait<'dynosaur_struct> {
        ptr: *mut (dyn ErasedMyTrait + 'dynosaur_struct),
        owned: bool,
    }
    impl<'dynosaur_struct> MyTrait for DynMyTrait<'dynosaur_struct> {
        fn foo(&self) -> impl Send {
            let fut: ::core::pin::Pin<Box<dyn Send + '_>> =
                unsafe { &*self.ptr }.foo();
            let fut: ::core::pin::Pin<Box<dyn Send + 'static>> =
                unsafe { ::core::mem::transmute(fut) };
            fut
        }
    }
    impl<'dynosaur_struct> DynMyTrait<'dynosaur_struct> {
        pub fn new<DYNOSAUR>(value: DYNOSAUR) -> Self where
            DYNOSAUR: MyTrait + 'dynosaur_struct {
            let value = Box::new(value);
            Self {
                ptr: Box::into_raw(value as
                            Box<dyn ErasedMyTrait + 'dynosaur_struct>) as
                    *mut (dyn ErasedMyTrait + 'dynosaur_struct),
                owned: true,
            }
        }
        pub fn from_ref<DYNOSAUR>(value: &'dynosaur_struct DYNOSAUR)
            -> ::dynosaur::macro_lib::Ref<Self> where DYNOSAUR: MyTrait +
            'dynosaur_struct {
            let this =
                Self {
                    ptr: value as
                                &'dynosaur_struct (dyn ErasedMyTrait + 'dynosaur_struct) as
                            *const (dyn ErasedMyTrait + 'dynosaur_struct) as
                        *mut (dyn ErasedMyTrait + 'dynosaur_struct),
                    owned: false,
                };
            unsafe { ::dynosaur::macro_lib::Ref::new(this) }
        }
        pub fn from_mut<DYNOSAUR>(value: &'dynosaur_struct mut DYNOSAUR)
            -> ::dynosaur::macro_lib::RefMut<Self> where DYNOSAUR: MyTrait +
            'dynosaur_struct {
            let this =
                Self {
                    ptr: value as
                            &'dynosaur_struct mut (dyn ErasedMyTrait + 'dynosaur_struct)
                        as *mut (dyn ErasedMyTrait + 'dynosaur_struct),
                    owned: false,
                };
            unsafe { ::dynosaur::macro_lib::RefMut::new(this) }
        }
    }
}
use _dynosaur_macro_DynMyTrait::DynMyTrait;

fn main() {}
